pages/docs/adapters/calculix/adapter-calculix-pastix-build.md

```diff
@@ -0,0 +1,166 @@
+- Some parts of the code require older versions of the GNU compilers. You may have to replace `gcc` by `gcc-7` and similarly for `g++` and `gfortran` in the `make_pastix.sh` script. This requires installing the relevant Ubuntu packages.
+
+## Building ARPACK, a CalculiX dependency
+
+Calculix relies on ARPACK, and when built with PaStiX, we cannot rely on standard distributions of that library, because it doesn't feature 8-bytes integers by default. We need to compile it ourself. The source code can be found [here](https://www.caam.rice.edu/software/ARPACK/), or by running these commands :
+
+```bash
+cd ~
+wget https://www.caam.rice.edu/software/ARPACK/SRC/arpack96.tar.gz
+wget https://www.caam.rice.edu/software/ARPACK/SRC/patch.tar.gz
+zcat arpack96.tar.gz | tar -xvf -
+zcat patch.tar.gz    | tar -xvf -
+
+```
+
+Before building the library, the following modifications are required :
+
+- In `ARmake.inc`, change `PLAT` by the appropriate suffix (the adapter's makefile assumes INTEL)
+- In `ARmake.inc`, change the Fortran compilation flags to add `-fdefault-integer-8`. You may also need to remove the flag `-cg89`.
+- If you extracted the archive on another folder than your home repository, update `home` in `ARmake.inc` accordingly.
+- In the file `UTIL/second.f`, comment with a star the line containing `EXTERNAL ETIME`.
+
+Once all of these are done, simply run `make lib` in the `ARPACK` folder.
+
+## Building the adapter
+
+To build the adapter, clone its repository and checkout the `2.17` branch :
+
+```bash
+    git clone -b v2.17 https://github.com/precice/calculix-adapter
+    cd calculix-adapter
+```
+
+### Fixing the code
+
+Due to some conflicts between CalculiX, PaStiX and the adapter (both CalculiX and PaStiX have a `pastix.h` file, and neither of them is local from the point of view of the adapter), some changes are required in the CalculiX codebase. We provide a script, `pastix_pre_build.sh` that does these changes. Run it.
+
+```bash
+    ./pastix_pre_build.sh
+```
+
+### Compilation
+
+To build the adapter, use the provided `Makefile_i8_PaStiX` : the regular Makefile would build the adapter without PaStiX. Assuming you followed the previous steps, it should be useable without modifications; otherwise, some paths updates could be required. Run this command in the `calculix-adapter` (and be sure to checkout the `2.17` branch) folder :
+
+```bash
+    make -f Makefile_i8_PaStiX -j 4 
+```
+
+Once the build is successful, the adapter should be in `./bin/ccx_preCICE`.
+
+### Updating shared libraries
+
+Running the adapter at this point should fail because of a missing shared library : `libparsec.so.2`. A possible fix to this is to copy it in your local library folder and run `ldconfig` :
+
+```bash
+    sudo cp ~/PaStiX/parsec_i8/lib/libparsec.so.2 /usr/local/lib
+    sudo ldconfig
```

Comment:
I again don't really like the system-wide paths.

Reply:
What's the alternative for shared libraries ? I'm used to paste DLL locally on Windows but AFAIK there's no equivalent on Linux.

Reply:
I would build in local directories and use, e.g., pkg-config. Maybe this page helps: https://precice.org/installation-linking.html

Reply:
So, I investigated the `ldconfig` sutff, and I found two cleaner approaches:

- You can create a file in /etc/ld.so.conf.d/ (e.g. `parsec.conf`) where you put the path of of the lib (`~/PaStiX/parsec_i8/lib/libparsec.so.2 /usr/local/lib`, except that the ~ alias must be replaced by its true value) then run `ldconfig`. So instead of copying the library we add an index to it in the global system
- You can simply `export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:path/to/lib` and it works without calling `ldconfig`. But we must make sure it's exported everytime, for instance by adding this line in a `.bashrc` file or something like this. It's probably better since no superuser call is required but I'm not sure it's the cleanest.

I read the page about pkg-config, but this is related to the issue of the makefile with absolute paths (which I've yet to improve), not to the `ldconfig` issue, right ? Not sure why you linked in a comment about `ldconfig`, unless I missed something.

Reply:
It is fine if you just use the `LD_LIBRARY_PATH` option, we anyway use it for other dependencies as well. This is very common and easy, even though not ideal. And yes, one would normally add the change in their `.bashrc`.

Reply:
> I read the page about pkg-config, but this is related to the issue of the makefile with absolute paths (which I've yet to improve), not to the ldconfig issue, right ?

One can use `pkg-config` similarly: you can make a `.pc` file describing the include and link flags, and then make this file discoverable, by extending the `PKG_CONFIG_PATH`. But this again requires changing an environment variable.

> Not sure why you linked in a comment about ldconfig, unless I missed something.

I am not sure I understand this part. My objection was mainly about `sudo`, which is needed if we want the user to install dependencies in system-wide paths.

Reply:
`LD_LIBRARY_PATH` it is then.

> One can use pkg-config similarly: you can make a .pc file describing the include and link flags, and then make this file discoverable, by extending the PKG_CONFIG_PATH. But this again requires changing an environment variable.

I updated the makefile to use `pkg-config` to find flags for `pastix hwloc parsec` and it works. So I add in the documentation that we must export these 3 paths, right ? (It's already in the Makefile comment)

> I am not sure I understand this part.

What I mean is that I changed two things:
- I update LD_CONFIG_PATH so that at runtime, the parsec `.so` file is found
- I improved the makefile to use `pkg-config` to find correct include / link flags for dependencies of the adapter instead of hardcoding them.

But I don't see how the second point removes a `sudo`.

> My objection was mainly about sudo, which is needed if we want the user to install dependencies in system-wide paths.

With these changes the only `sudo` left are the one for installing dependencies (apt install) and for installing OpenBLAS system wide. I guess I should try to make it possible to install OpenBLAS locally, but for some reasons I had troubles the first time I tried. I'll give it another try !

Reply:
I guess this conversation was evolved for a while and we have lost synch a bit (and potentially me to myself). Let me just summarize my concerns and I believe you will know what exactly to do afterwards:

- Let's ask the user explicitly to install dependencies, instead of running a script that installs them (anyway ok)
- Dependencies from the package manager are better than installing them manually (if we can assume a system). In that case, `sudo` is perfectly fine.
- Additional information on building the dependencies from source as a fallback is welcome (for clusters etc), but optional.
- Only ask the user to build dependencies from source by default if they need some special option or any other modifications.
- Avoid running `sudo make install` or modifying system paths, even if this is `/usr/local`. This is allowed (and this is the purpose of `/usr/local`), but users often mess up their systems if they manually install software on system paths (e.g., by having duplicate installations). Asking them to install to a local prefix gives additional confidence that they can easily fix it later.